http-transport-binding.md

```diff
@@ -226,7 +226,10 @@ with [RFC7230, sections 3, 3.2, 3.2.6][rfc7230-section-3]. The rules for
 encoding of the percent character ('%') apply as defined in [RFC 3986 Section
 2.4.][rfc3986-section-2-4].
 
-JSON objects and arrays are NOT surrounded with single or double quotes.
+[JSON values][json-value] for well-known context attributes as defined by the
+[CloudEvents specification][ce] or [CloudEvents Extensions][extensions] MAY omit
+the surrounding single or double quote. All values are assumed to be a string
+unless well-known.
```

Comment:
I'm not sure if I get this correctly - if I send a not well-known attribute of type `Integer` or `Binary` over HTTP Binary, a valid implementation of the HTTP Binary Transport MUST transform the attribute into a `String`?

Reply:
I wonder if it wouldn't be better to not talk about "well-known" attributes since that implies spec defined attributes would be treated differently from extensions, but instead talk about the data type of the attribute. Meaning, for types like 'string' we remove the quotes.  For integers, we're kind of stuck when extensions are copied from http headers into whatever CE representation is being used.

Another option is to make all attributes in CE variants of strings. Then if someone wants to convert it to some other datatype when they extract it from the in-memory representation of the CE, that's up to them.

Reply:
>  a valid implementation of the HTTP Binary Transport MUST transform the attribute into a String

It always sends it as a string in the header. We are currently pretending that a string is an escaped byte blob and that is a really strange choice for http headers.

Reply:
> implies spec defined attributes would be treated differently from extensions

I am saying exactly that. The exception is for known extensions.

Reply:
"Well-known" is difficult because that set might change in CloudEvents v.Next and it might include promotion of previously proprietary attributes into the standard. For proprietary attributes you can't omit the quotes, but for well-known ones you can and thus the promotion of the attribute might break existing code. I understand that we could solve that through versioning, but I consider that a dirty trick.

There are three levels of strings here:
1. The string that holds the JSON data and which is stored in the HTTP header
2. The JSON string object that holds string data
3. The string data 

The quotes belong to 2. If you parse the string, you first get a JSON object for which you then get inferred type info and cast it to a JSON string and then you convert to a platform string.

Reply:
> It always sends it as a string in the header. We are currently pretending that a string is an escaped byte blob and that is a really strange choice for http headers.

Yes, I'm aware that at HTTP level, we're always sending strings.

But I am talking about the CloudEvents type system when the event is transformed out of HTTP into something else (e.g. an in-memory representation of an SDK).
Maybe I'm wrong, but - as far as I understand your proposed text - if I'm sending an attribute `myintegerattr` of type `Integer`, it'll come back as type `String`. In C-like syntax, `int myintegerattr` will come back as `String myintegerattr`.
So the type information gets lost, and the event does not conform to the types a consumer may expect. Or, put another way, if I send the same event in Binary and Structured mode, I get different events back.

Reply:
Yup. This difference in how we deal with things like Integers, combined with our inability to promote extensions in a dot-release leads, me to think that CloudEvents should treat everything as strings - at the spec level. What any particular implementation does with those values, and how they expose them to their users, is an implementation choice.

Of course, that will mean that JSON CEs will now have integers quoted, but TBH that's less weird than quoted HTTP Header strings.

Reply:
A simple out for this problem is to wrap the value:

Binary to Structured:  

`ce-custom-num: 1` converts to`{"custom":{"num":"1"}}` . <-- this is bad. 

`ce-custom: {"num":1}` converts to`{"custom":{"num":1}}` . <-- this is good.

I think it is an easy way to get the best of both worlds. Don't flatten the map. Don't support Strings at the root level of extensions.

Reply:
> Of course, that will mean that JSON CEs will now have integers quoted, but TBH that's less weird than quoted HTTP Header strings.

Yes, it is a tradeoff decision. But we should also consider the likelihood of one format being used over the other. E.g. if 50% of the future users will use the Binary encoding, then it makes sense to optimize for a less-weird Binary encoding. However, if only 5% use it, and the other 95% use formats that properly support Integers (correct me if I'm wrong, but that seems to be all others formats?), then I'd rather have a weird design choice for the 5%, and a non-weird design choice for the 95%.

At this point it is mostly guesswork, but personally I'd not optimize for HTTP Binary, because I don't think it'll be the major way that CloudEvents will be transported.

Reply:
Knative only supports Binary mode fwiw. The reason being that the consumer can still decode the payload in the same way it did before.